### PR TITLE
Support TLS on TCP servers

### LINKED ISSUE #68: Add TLS Support
For the TCP transport, we should look at supporting TLS. As a starting point, there are [TLS APIs for `tokio`](https://github.com/tokio-rs/tls), which is the async runtime this server uses.

### PR DESCRIPTION
**Pull request by KodrAus**
Closes #68 

This PR adds support for secure streams when using GELF's TCP protocol through the [rustls](https://docs.rs/rustls/latest/rustls/) library. It supports certificates in the [PEM](https://docs.rs/rustls-pemfile/latest/rustls_pemfile/) container format.

A lot of the async code here predates `async`/`await` support in the language, so I'm updating bits and pieces of it conservatively as I need to touch them. TLS support isn't too intrusive though, I think it works out pretty nicely.

### CODE REVIEW

File: seq-input-gelf.d.json [hunk 1]
```diff
@@ -20,6 +20,16 @@
         "helpText": "The url (protocol, IP address and port) on which the input will listen for GELF payloads. The default is `udp://0.0.0.0:12201`. Specifying the protocol as `tcp` will listen on TCP instead of UDP.",
         "isOptional": true
       },
+      "certificatePath": {
+        "displayName": "Certificate path",
+        "helpText": "The absolute path to a `.pem` certificate file to use for secure transport. This setting is only used when the protocol is TCP.",
+        "isOptional": true
+      },
+      "certificatePasswordPath": {
+        "displayName": "Certificate password path",
```
Comment (nblumhardt) [inline seq-input-gelf.d.json:29]: "Certificate private key path"?

Comment (nblumhardt) [reply seq-input-gelf.d.json:29]: If this setting is provided, TLS will be forced

File: sqelf/src/server/mod.rs [hunk 2]
```diff
@@ -177,9 +181,37 @@ pub fn build(
                 Either::Left(server)
             }
             Protocol::Tcp => {
+                let tls_config = if let Some(Certificate {
```
Comment (KodrAus) [inline sqelf/src/server/mod.rs:184]: Error if the protocol is UDP and we have certificate config

File: sqelf/src/server/mod.rs [hunk 3]
```diff
@@ -177,9 +181,37 @@ pub fn build(
...
+                        .collect();
+
+                    let mut reader = BufReader::new(File::open(password_path)?);
+                    let mut keys = rustls_pemfile::pkcs8_private_keys(&mut reader)?;
+
+                    let config = rustls::ServerConfig::builder()
+                        .with_safe_defaults()
+                        .with_no_client_auth()
+                        .with_single_cert(cert, rustls::PrivateKey(keys.remove(0)))
+                        .unwrap();
```
Comment (KodrAus) [inline sqelf/src/server/mod.rs:204]: Error rather than `unwrap`

**Review by nblumhardt — APPROVED** [review]

**Pull request merged**